src/appshell/view/dockwindow/dockpageview.cpp

```diff
@@ -179,22 +179,36 @@ bool DockPageView::isDockOpen(const QString& dockName) const
 void DockPageView::toggleDock(const QString& dockName)
 {
     DockBase* dock = dockByName(dockName);
+    if (dock == nullptr) {
+        return;
+    }
+
+    bool shouldOpen = true;
+
     DockPanelView* panel = dynamic_cast<DockPanelView*>(dock);
     if (panel) {
-        DockPanelView* destinationPanel = findPanelForTab(panel);
-        if (destinationPanel) {
-            const int currentTabIndex = destinationPanel->currentTabIndex();
-            const int tabIndex = destinationPanel->tabIndexOfPanel(panel);
-            const bool tabNotFound = tabIndex < 0;
-            const bool tabNotCurrent = tabIndex != currentTabIndex;
-            const bool openTab = tabNotFound || tabNotCurrent;
-            setDockOpen(dockName, openTab);
-
-            return;
+        if (panel->isOpen()) {
+            if (panel->floating()) {
+                shouldOpen = !panel->hasFocus();
```

Comment:
The calls to hasFocus() are to meet the requirement that the hotkey should first focus the panel (if it's open and not focused), and only then is it ok to close it.